core/src/main/java/org/apache/iceberg/expressions/ExpressionParser.java

```diff
@@ -0,0 +1,334 @@
+    generator.writeStringField(TYPE, NAMED_REFERENCE);
+    generator.writeStringField(VALUE, term.name());
+    generator.writeEndObject();
+  }
+
+  public static void toJson(List<Literal<?>> literals, JsonGenerator generator) throws IOException {
+    generator.writeStartArray();
+    for (int i = 0; i < literals.size(); i++) {
+      toJson(literals.get(i), generator);
+    }
+    generator.writeEndArray();
+  }
+
+  public static void toJson(Literal<?> literal, JsonGenerator generator) throws IOException {
+    generator.writeStartObject();
+
+    Object value = literal.value();
+    Type type;
+    if (value instanceof Boolean) {
+      type = Types.BooleanType.get();
+    } else if (value instanceof Integer) {
+      type = Types.IntegerType.get();
+    } else if (value instanceof Long) {
+      type = Types.LongType.get();
+    } else if (value instanceof Float) {
+      type = Types.FloatType.get();
+    } else if (value instanceof Double) {
+      type = Types.DoubleType.get();
+    } else if (value instanceof CharSequence) {
+      type = Types.StringType.get();
+    } else if (value instanceof UUID) {
+      type = Types.UUIDType.get();
+    } else if (value instanceof byte[]) {
+      type = Types.FixedType.ofLength(((byte[]) value).length);
+    } else if (value instanceof ByteBuffer) {
+      if (literal instanceof Literals.FixedLiteral) {
+        type = Types.FixedType.ofLength(((ByteBuffer) value).remaining());
+      } else {
+        type = Types.BinaryType.get();
+      }
+    } else if (value instanceof BigDecimal) {
+      BigDecimal decimal = (BigDecimal) value;
+      type = Types.DecimalType.of(decimal.precision(), decimal.scale());
+    } else {
+      throw new IllegalArgumentException("Cannot find literal type for value class " + value.getClass().getName());
+    }
+
+    generator.writeStringField(TYPE, type.toString());
+    generator.writeStringField(VALUE, StandardCharsets.UTF_8.decode(literal.toByteBuffer()).toString());
+    generator.writeEndObject();
+  }
+
+  public static Expression fromJson(String json) {
+    try {
+      ObjectMapper mapper = new ObjectMapper();
+      return fromJson(mapper.readTree(json));
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    }
+  }
+
+  public static Expression fromJson(JsonNode json) {
+    String expressionType = JsonUtil.getString(TYPE, json);
+
+    if (AND.equals(expressionType)) {
+      return new And(fromJson(json.get(LEFT_OPERAND)), fromJson(json.get(RIGHT_OPERAND)));
+    } else if (OR.equals(expressionType)) {
+      return new Or(fromJson(json.get(LEFT_OPERAND)), fromJson(json.get(RIGHT_OPERAND)));
+    } else if (NOT.equals(expressionType)) {
+      return new Not(fromJson(json.get(OPERAND)));
+    } else if (TRUE.equals(expressionType)) {
+      return True.INSTANCE;
+    } else if (FALSE.equals(expressionType)) {
+      return False.INSTANCE;
+    } else if (PREDICATE_TYPES.contains(expressionType)) {
```

Comment:
I think we don't need a set of predicate types, we just need to have a single else block, delegating all the rest of the conditions to `fromJsonToPredicate(json, expressionType);`